lib/dao.js

```diff
@@ -1526,6 +1526,27 @@ function NumberType(val) {
   return !isNaN(num) ? num : val;
 }
 
+function coerceArray(val) {
+  if (Array.isArray(val)) {
+    return val;
+  }
+
+  if (!utils.isPlainObject(val)) {
+    throw new Error(g.f('Value is not an {{array}} or {{array-like object}}'));
+  }
+
+  var arrayVal = new Array(Object.keys(val).length);
+  for (var i = 0; i < arrayVal.length; ++i) {
+    if (!val.hasOwnProperty(i)) {
+      throw new Error(g.f('Value is not an {{array}} or {{array-like object}}'));
```

Comment:
If I understand this code (and the test cases) correctly, when processing my example `result[4000000000]=true`, you will only access `result[0]` and bail out with an error.

I am wondering whether we should iterate over all keys in the input `val`, thus converting `result[4000000000]=true` to a single-item array `[true]`?

```js
var arrayVal = [];
for (var key in val) {
  arrayVal.push(val[key]);
}
return arrayVal;
```

If we want to reject such requests, then I would like to see a better error message reported back to the user. Right now, `coerceArray({ 40000: true })` fails with `Value is not an array or array-like object`, which is misleading to me, because the input (`result[40000]=true`) is an array-like object.

@doublemarked Thoughts?